SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp

```diff
@@ -739,47 +764,69 @@ int64 USpatialActorChannel::ReplicateActor()
 	if (NetDriver->LoadBalanceStrategy != nullptr &&
 		NetDriver->StaticComponentView->HasAuthority(EntityId, SpatialConstants::AUTHORITY_INTENT_COMPONENT_ID))
 	{
-		if (!NetDriver->LoadBalanceStrategy->ShouldHaveAuthority(*Actor) && !NetDriver->LockingPolicy->IsLocked(Actor))
+		if (MigrationVirtualWorkerId.IsSet())
 		{
-			const AActor* NetOwner = Actor->GetNetOwner();
+			AActor* HierarchyRoot = SpatialGDK::GetHierarchyRoot(Actor);
 
-			uint64 HierarchyAuthorityReceivedTimestamp = AuthorityReceivedTimestamp;
-			if (NetOwner != nullptr)
+			if (NetDriver->LockingPolicy->IsLocked(Actor))
 			{
-				GetLatestAuthorityChangeFromHierarchy(NetOwner, HierarchyAuthorityReceivedTimestamp);
-			}
-
-			const float TimeSinceReceivingAuthInSeconds = double(FPlatformTime::Cycles64() - HierarchyAuthorityReceivedTimestamp) * FPlatformTime::GetSecondsPerCycle64();
-			const float MigrationBackoffTimeInSeconds = 1.0f;
-
-			if (TimeSinceReceivingAuthInSeconds < MigrationBackoffTimeInSeconds)
-			{
-				UE_LOG(LogSpatialActorChannel, Verbose, TEXT("Tried to change auth too early for actor %s"), *Actor->GetName());
+				DropActorHierarchyMigration(NetDriver, HierarchyRoot);
 			}
-			else
+			else if (MigrationFrame != NetDriver->ReplicationFrame)
 			{
-				const VirtualWorkerId NewAuthVirtualWorkerId = NetDriver->LoadBalanceStrategy->WhoShouldHaveAuthority(*Actor);
-				if (NewAuthVirtualWorkerId == SpatialConstants::INVALID_VIRTUAL_WORKER_ID)
+				if (MigrationVirtualWorkerId.GetValue() == SpatialConstants::INVALID_VIRTUAL_WORKER_ID)
 				{
-					UE_LOG(LogSpatialActorChannel, Error, TEXT("Load Balancing Strategy returned invalid virtual worker for actor %s"), *Actor->GetName());
+					const VirtualWorkerId NewAuthVirtualWorkerId = NetDriver->LoadBalanceStrategy->WhoShouldHaveAuthority(*HierarchyRoot);
+					if (NewAuthVirtualWorkerId == SpatialConstants::INVALID_VIRTUAL_WORKER_ID)
+					{
+						UE_LOG(LogSpatialActorChannel, Error, TEXT("Load Balancing Strategy returned invalid virtual worker for actor %s"), *Actor->GetName());
+						DropActorHierarchyMigration(NetDriver, HierarchyRoot);
+					}
+					else if (NewAuthVirtualWorkerId == NetDriver->LoadBalanceStrategy->GetLocalVirtualWorkerId())
+					{
+						DropActorHierarchyMigration(NetDriver, HierarchyRoot);
+					}
```

Comment:
Why do you need these drop hierarchy calls? If the above condition `MigrationVirtualWorkerId.GetValue() == SpatialConstants::INVALID_VIRTUAL_WORKER_ID` is true then we know the whole hierarchy has the same set?

Reply:
So at first, I am marking the hierarchy as needing to be migrated by initializing the TOptional with an invalid worker Id. On the next frame, if the optional is valued and we have an invalid worker Id, it means we have yet to determine the destination, and have encountered the first actor in the hierarchy.

We actually computing the destination, and if it results in an invalid worker Id, or the local worker Id, I drop the attempt to migrate for the whole hierarchy for this frame (no need for the other actors to try the same computation).

Reply:
Ah okay, I see, thanks. Do `MigrationFrame != NetDriver->ReplicationFrame` and `MigrationVirtualWorkerId.IsSet()` represent the same state then? I.e. we change the top level conditional for the section to `MigrationFrame != NetDriver->ReplicationFrame` then MigrationVirtualWorkerId doesn't need to be optional?

Reply:
They are not the same state.
MigrationWorkerVirtualId.IsSet() means "someone in the hierarchy has requested a migration on this frame, or on a previous frame".
MigrationFrame != NetDriver->ReplicationFrame means that we are at a different replication frame than the one where we determined we should migrate.

Reply:
My point is that you can replace `MigrationWorkerVirtualId.IsSet()` with  `MigrationFrame != NetDriver->ReplicationFrame` on the top level conditional and make `MigrationWorkerVirtualId` not be an optional anymore?

Reply:
I think this conversation is hidden cos outdated so moving to the correct place